depends/libhdfs3/src/client/CryptoCodec.cpp

```
@@ -25,154 +25,188 @@
+		char IV[initIV.length()];
+
+		int i = initIV.length(); // IV length
+		int j = 0; // counter bytes index
+		unsigned int sum = 0;
+		while (i-- > 0) {
+			// (sum >>> Byte.SIZE) is the carry for addition
+			sum = (initIV[i] & 0xff) + (sum >> 8);
+			if (j++ < 8) { // Big-endian, and long is 8 bytes length
+				sum += (char) counter & 0xff;
+				counter >>= 8;
+			}
+			IV[i] = (char) sum;
+		}
+
+		return std::string(IV, initIV.length());
+	}
+
+	CryptoCodec::CryptoCodec(FileEncryptionInfo *encryptionInfo, shared_ptr<KmsClientProvider> kcp, int32_t bufSize) :
+		encryptionInfo(encryptionInfo), kcp(kcp), bufSize(bufSize)
+	{
+
+		// Init global status
+		ERR_load_crypto_strings();
+		OpenSSL_add_all_algorithms();
+		OPENSSL_config(NULL);
+
+		// Create cipher context
+		cipherCtx = EVP_CIPHER_CTX_new();
+		cipher = NULL;
+
+		padding = 0;
+		counter = 0;
+		is_init = false;
+	}
+
+	CryptoCodec::~CryptoCodec()
+	{
+		if (cipherCtx)
+			EVP_CIPHER_CTX_free(cipherCtx);
+	}
+
+	std::string CryptoCodec::getDecryptedKeyFromKms()
+	{
+		ptree map = kcp->decryptEncryptedKey(*encryptionInfo);
+		std::string key;
+		try {
+			key = map.get < std::string > ("material");
+		} catch (...) {
+			THROW(HdfsIOException, "CryptoCodec : Can not get key from kms.");
+		}
+
+		int rem = key.length() % 4;
+		if (rem) {
+			rem = 4 - rem;
+			while (rem != 0) {
+				key = key + "=";
+				rem--;
+			}
+		}
+
+		std::replace(key.begin(), key.end(), '-', '+');
+		std::replace(key.begin(), key.end(), '_', '/');
+
+		LOG(INFO, "CryptoCodec : getDecryptedKeyFromKms material is :%s", key.c_str());
+
+		key = KmsClientProvider::base64Decode(key);
+		return key;
+	}
+
+	int CryptoCodec::init(CryptoMethod crypto_method, int64_t stream_offset) {
+		//check already init
+		if (is_init)
+			return 0;
+
+		// Get decrypted key from KMS
+		std::string key;
+		try {
+			key = getDecryptedKeyFromKms();
+		}
+		catch (...) {
+			LOG(WARNING, "getDecryptedKeyFromKms failed");
```

It's better to throw the exception to upper function then return error.